core/pta/attestation.c

```diff
@@ -0,0 +1,533 @@
+	size_t read_len = 0;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_pobj_get(uuid, (void *)obj_id, obj_id_len, flags, false, fops,
+			   &po);
+	if (res)
+		return res;
+
+	res = po->fops->open(po, &file_size, &fh);
+	if (res)
+		goto out;
+
+	read_len = *len;
+	res = po->fops->read(fh, offset, data, &read_len);
+	if (res == TEE_ERROR_CORRUPT_OBJECT) {
+		EMSG("Object corrupt");
+		po->fops->remove(po);
+	} else if (!res) {
+		*len = read_len;
+	}
+
+	po->fops->close(&fh);
+out:
+	tee_pobj_release(po);
+
+	return res;
+}
+
+static TEE_Result sec_storage_obj_write(TEE_UUID *uuid, uint32_t storage_id,
+					const uint8_t *obj_id,
+					size_t obj_id_len,
+					const uint8_t *data, size_t len,
+					size_t offset, uint32_t flags)
+
+{
+	const struct tee_file_operations *fops = NULL;
+	struct tee_file_handle *fh = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	struct tee_pobj *po = NULL;
+
+	fops = tee_svc_storage_file_ops(storage_id);
+	if (!fops)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (obj_id_len > TEE_OBJECT_ID_MAX_LEN)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = tee_pobj_get(uuid, (void *)obj_id, obj_id_len, flags, false,
+			   fops, &po);
+	if (res)
+		return res;
+
+	res = po->fops->open(po, NULL, &fh);
+	if (res == TEE_ERROR_ITEM_NOT_FOUND)
+		res = po->fops->create(po, false, NULL, 0, NULL, 0, NULL, 0,
+				       &fh);
+	if (!res) {
+		res = po->fops->write(fh, offset, data, len);
+		po->fops->close(&fh);
+	}
+
+	tee_pobj_release(po);
+
+	return res;
+}
+
+static TEE_Result load_key(void)
+{
+	/*
+	 * Serialized key pair is 3 bignums (e, p and n) plus their sizes
+	 * encoded as uint16_t. e is 65537 so it needs only 3 bytes.
+	 */
+	size_t size = 3 + 2 * MAX_KEY_SIZE / 8 + 3 * sizeof(uint16_t);
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *buf;
+
+	buf = calloc(size, 1);
+	if (!buf)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	DMSG("Loading RSA key pair from secure storage");
+	res = sec_storage_obj_read(&pta_uuid, TEE_STORAGE_PRIVATE,
+				   key_file_name, sizeof(key_file_name) - 1,
+				   buf, &size, 0, TEE_DATA_FLAG_ACCESS_READ);
+	if (res)
+		goto out;
+	DMSG("Read %zu bytes", size);
+	res = deserialize_key(buf, size);
+	if (!res)
+		DMSG("Loaded %zu bit key pair", crypto_bignum_num_bits(key->n));
+
+out:
+	free(buf);
+	return res;
+}
+
+static TEE_Result write_key(void)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint8_t *buf = NULL;
+	size_t size = 0;
+
+	DMSG("Saving key pair");
+	res = serialize_key(&buf, &size);
+	if (res)
+		return res;
+
+	res = sec_storage_obj_write(&pta_uuid, TEE_STORAGE_PRIVATE,
+				    key_file_name, sizeof(key_file_name) - 1,
+				    buf, size, 0, TEE_DATA_FLAG_ACCESS_WRITE);
+	if (!res)
+		DMSG("Wrote %zu bytes", size);
+	free(buf);
+	return res;
+}
+
+static TEE_Result init_key(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+
+	if (!key) {
+		res = load_key();
```

Comment:
Agreed, some flexibility can easily be added for instance with weak functions. I'd rather wait until a use case is clarified so that we can define a suitable interface.